lighthouse-core/gather/gatherers/styles.js

```diff
@@ -0,0 +1,144 @@
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview Gathers the active style and stylesheets used on a page.
+ * "Active" means that if the stylesheet is removed at a later time
+ * (before endStylesCollect is called), this gatherer will not include it.
+ */
+
+'use strict';
+
+const WebInspector = require('../../lib/web-inspector');
+const Gatherer = require('./gatherer');
+
+/**
+ * @param {!gonzales.AST} parseTree
+ * @return {!Array}
+ */
+function getCSSPropsInStyleSheet(parseTree) {
+  let results = [];
+
+  parseTree.traverseByType('declaration', function(node, index, parent) {
+    const keyVal = node.toString().split(':').map(item => item.trim());
+    results.push({
+      property: {name: keyVal[0], val: keyVal[1]},
+      declarationRange: node.declarationRange,
+      selector: parent.selectors.toString()
+    });
+  });
+
+  return results;
+}
+
+class Styles extends Gatherer {
+
+  constructor() {
+    super();
+    this._activeStyleSheetIds = [];
+    this._activeStyleHeaders = {};
+    this._onStyleSheetAdded = this.onStyleSheetAdded.bind(this);
+    this._onStyleSheetRemoved = this.onStyleSheetRemoved.bind(this);
+  }
+
+  onStyleSheetAdded(styleHeader) {
+    // Exclude stylesheets "injected" by extensions or ones that were added by
+    // users using the "inspector".
+    if (styleHeader.header.origin !== 'regular') {
+      return;
+    }
+
+    this._activeStyleHeaders[styleHeader.header.styleSheetId] = styleHeader;
+    this._activeStyleSheetIds.push(styleHeader.header.styleSheetId);
+  }
+
+  onStyleSheetRemoved(styleHeader) {
+    delete this._activeStyleHeaders[styleHeader.styleSheetId];
+
+    const idx = this._activeStyleSheetIds.indexOf(styleHeader.styleSheetId);
+    if (idx !== -1) {
+      this._activeStyleSheetIds.splice(idx, 1);
+    }
+  }
+
+  beginStylesCollect(driver) {
+    driver.sendCommand('DOM.enable');
+    driver.sendCommand('CSS.enable');
+    driver.on('CSS.styleSheetAdded', this._onStyleSheetAdded);
+    driver.on('CSS.styleSheetRemoved', this._onStyleSheetRemoved);
+  }
+
+  endStylesCollect(driver) {
+    return new Promise((resolve, reject) => {
+      if (!driver || !this._activeStyleSheetIds.length) {
+        reject('No active stylesheets were collected.');
+        return;
+      }
+
+      const parser = new WebInspector.SCSSParser();
+
+      // Get text content of each style.
+      const contentPromises = this._activeStyleSheetIds.map(sheetId => {
+        return driver.sendCommand('CSS.getStyleSheetText', {
+          styleSheetId: sheetId
+        }).then(content => {
+          const styleHeader = this._activeStyleHeaders[sheetId];
+          styleHeader.content = content.text;
+          styleHeader.parsedContent = getCSSPropsInStyleSheet(
+              parser.parse(styleHeader.content, {syntax: 'css'}));
+          return content;
+        });
+      });
+
+      Promise.all(contentPromises).then(_ => {
```

Comment:
the results of `contentPromises` is the input to this `then` function, so no need to reconstruct from `Object.keys(this._activeStyleHeaders)`. Also lets you bring in `sendCommand` calls. Something like

```js
Promise.all(contentPromises).then(stylesheets => {
  driver.off('CSS.styleSheetAdded', this._onStyleSheetAdded);
  driver.off('CSS.styleSheetRemoved', this._onStyleSheetRemoved);
  return driver.sendCommand('CSS.disable')
    .then(_ => driver.sendCommand('DOM.disable'))
    .then(_ => resolve(stylesheets));
}).catch(err => reject(err));
```
(or one of the many promise chaining permutations that could be done with this)